ArduCopter/mode_auto.cpp

```diff
@@ -771,33 +771,99 @@ void Copter::ModeAuto::wp_run()
+            // get pilot's desired yaw rate
+            target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+            if (!is_zero(target_yaw_rate)) {
+                auto_yaw.set_mode(AUTO_YAW_HOLD);
+            }
+        }
+    }
+
+   // set motors to full range
     motors->set_desired_spool_state(AP_Motors::DESIRED_THROTTLE_UNLIMITED);
 
-    // run waypoint controller
-    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());
+    // run waypoint controller 
+    copter.failsafe_terrain_set_status(wp_nav->update_wpnav());  
+      
+     // mode semi auto
+    if (g.auto_man_alt == 1) {
+        
+        // get pilot desired climb rate alt control Alt_Hold add a2sAndres
+        float target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+        target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
+
+        // get pilot desired lean angles Althold add a2sAndres
+        float target_roll, target_pitch;
+        get_pilot_desired_lean_angles(target_roll, target_pitch, copter.aparm.angle_max, attitude_control->get_althold_lean_angle_max());
+
+        //added control to resume flight add a2sAndres
+        if (target_roll != 0 || target_pitch != 0) {
+            //Change of mode to be able to restart the automatic mode
+            //Restart the automatic mode
+            copter.mode_auto.init(true);
+            //Re initialise wpnav targets after stopping when giving control of alt
+            wp_nav->shift_wp_origin_to_current_pos();
+        }
+        
+        // call attitude controller
+        // attitude Alt_Hold add a2sAndres
+        pos_control->set_speed_z(-get_pilot_speed_dn(), g.pilot_speed_up);
+        pos_control->set_accel_z(g.pilot_accel_z);
+        pos_control->set_alt_target_to_current_alt();
+        // change (inertial_nav.get_velocity_z ()) by (target_climb_rate) to add more control to the alt add a2sAndres
+        pos_control->set_desired_velocity_z(target_climb_rate);
+
+        // call position controller Alt_Hold add a2sAndres
+        target_climb_rate = get_pilot_desired_climb_rate(channel_throttle->get_control_in());
+        target_climb_rate = constrain_float(target_climb_rate, -get_pilot_speed_dn(), g.pilot_speed_up);
+
+        // adjust climb rate using rangefinder Alt_Hold add a2sAndres
+        if (copter.rangefinder_alt_ok()) {
+            // if rangefinder is ok, use surface tracking
+            target_climb_rate = get_surface_tracking_climb_rate(target_climb_rate, pos_control->get_alt_target(), G_Dt);
+        }
 
+        // update altitude target and call position controller Alt_Hold add a2sAndres
+        pos_control->set_alt_target_from_climb_rate_ff(target_climb_rate, G_Dt, false);
+    }
+        
     // call z-axis position controller (wpnav should have already updated it's alt target)
-    pos_control->update_z_controller();
+    pos_control->update_z_controller();    
```

Comment:
Wrong change, whitespaces added at the end.

Reply:
revised and changed